pkg/sql/vecindex/high_speed_backfill_design.md

```diff
@@ -0,0 +1,1102 @@
+
+#### Vector-Specific Logic Isolation
+Only vector indexes follow the new `STRUCTURE_BUILDING` → `STRUCTURE_BUILT` path. Regular secondary indexes bypass these phases entirely, maintaining full backward compatibility.
+
+### C-SPANN K-Means Integration
+
+The structure building algorithm leverages existing, proven components from the `pkg/sql/vecindex/cspann` package:
+
+#### Reused Components
+- **BalancedKmeans**: Binary k-means implementation adapted for hierarchical partition construction
+- **Workspace Management**: Existing memory allocation system for temporary vectors and computations  
+- **Quantizer Infrastructure**: Proven quantization algorithms (RaBitQ, etc.) for vector encoding
+- **Distance Metrics**: Support for L2Squared, Cosine, and InnerProduct distance functions
+- **Partition Structures**: Existing PartitionKey, TreeKey, and metadata infrastructure
+
+#### New Adaptations
+- **BulkKmeansBuilder**: Adapter that coordinates bulk construction while delegating to existing components
+- **Random Sampling**: Implementation of the `randint() % 100 == 0` sampling strategy for initial centroids
+- **Batch Processing**: Memory-bounded batch processing system with LRU caching and disk spilling
+- **Convergence Detection**: Monitoring system for k-means iteration termination
+
+### Backfill Infrastructure Integration
+
+The pre-built structure enables vector indexes to use standard bulk loading infrastructure without modification:
+
+#### Elimination of Vector-Specific Backfill Code
+- **No VectorIndexHelper.ReEncodeVector**: This function is completely removed from the backfill process
+- **Standard Chunk Construction**: Vector indexes use the same chunk building process as regular secondary indexes
+- **Bulk Adder Compatibility**: Pre-built structure allows direct use of high-performance bulk adders
+
+#### Temporary Index Structure Sharing
+- **Shared Partition Metadata**: Temporary indexes reference the main index's partition structure directly
+- **Reduced Write Operations**: Eliminates the need to copy partition metadata during merge operations
+- **Memory Efficiency**: Single structure serves both temporary and main indexes
+
+#### Fixup Process Integration
+- **Automatic Avoidance**: Existing fixup logic is modified to skip indexes under construction (non-PUBLIC status)
+- **No Explicit Disabling**: No need for special fixup disable/enable operations
+- **Natural Integration**: Works with existing `fixupProcessor.Enqueue()` logic
+
+### Performance Benefits
+
+#### Bulk Adder Performance
+The pre-built partition structure enables vector indexes to use the same high-performance bulk adders that other secondary indexes use, providing orders of magnitude performance improvements over transactional inserts.
```

Comment:
What's the impact of having the metadata records already pre-inserted? Will that interfere with the bulk adder, since I think those use `AddSSTable`?